components/controller/controller.js

```diff
@@ -0,0 +1,67 @@
+module.exports = () => {
+  const start = async ({
+    logger, metrics, bus, payment, slack,
+  }) => {
+    const { subscribe: initSubscription, publish: initPublication } = bus;
+    const subscribe = initSubscription(logger.error, logger.warn);
+    const publishPayoutTransformed = initPublication('payoutTransformed');
+    const publishPayoutFailedToTransform = initPublication('payoutFailedToTransform');
+
+    const trackMetric = event => metrics.trackMetric({ name: event, value: 1 });
+
+    const getChargedTransactions = transactions => transactions.filter(({ source }) => source.object === 'charge');
+
+    const handlePayout = shop => async message => {
+      const { properties: { messageId }, body: payout } = message;
+
+      const {
+        type,
+        data: {
+          object: {
+            id: payoutId,
+          },
+        },
+      } = payout;
+
+      const isPaid = event => event === 'payout.paid';
+
+      if (!isPaid(type)) {
+        logger.warn(`messageId ${messageId} | Skipping unpaid payout (event received: ${type}) | shop ${shop} | payout id ${payoutId}`);
+        trackMetric('payout_skipped');
+        return;
+      }
+
+      try {
+        const transactions = await payment.getBalanceTransactionsByPayout(shop, payoutId);
+        logger.info(`messageId ${messageId} | Fetched ${transactions ? transactions.length : 0} transactions for the payout ${payoutId}`);
+
+        if (transactions) {
+          const chargedTransactions = getChargedTransactions(transactions);
+          logger.info(`messageId ${messageId} | Found ${chargedTransactions ? chargedTransactions.length : 0} charge transactions | shop ${shop} | payoutId ${payoutId}`);
+        }
+
+        const payoutTransformed = { ...payout, transactions };
+        await publishPayoutTransformed(payoutTransformed);
```

Comment:
I'd add the compression. Payouts can potentially be big

```suggestion
        await publishPayoutTransformed(payoutTransformed, { contentEncoding: 'zlib' });
```

Reply:
+1